pywbem_mock/_wbemconnection_mock.py

```diff
@@ -995,6 +1012,104 @@ def _display_objects(obj_type, object_repo, namespace, cmt_begin, cmt_end,
+            raise ValueError("{0!A} is not a valid provider type. "
+                             "Valid provider types are {1!A}.", provider_type,
+                             provider_types)
+
+        if not isinstance(provider, DefaultInstanceProvider):
+            raise ValueError(
+                _format("{0!A} is not a valid subclass of "
+                        "DefaultInstanceProvider. ", provider))
+
+        if classnames is None:
+            raise ValueError(
+                _format('register_provider classnames argument '
+                        'must be string or list '
+                        'of strings. None not allowed.'))
+
+        if namespaces is not None:
+            if not isinstance(namespaces, (list, tuple, six.string_types)):
+                raise ValueError(
+                    _format('register_provider namespaces argument '
+                            '"{0!A}" not list or string'
+                            'of strings.', namespaces))
+
+            if isinstance(namespaces, six.string_types):
+                namespaces = [namespaces]
+
+            for namespace in namespaces:
+                if namespace not in self.namespaces:
+                    raise ValueError(
+                        _format("register_provider namespaces argument "
+                                '"{0!A}" does not exist in {1!A} '
+                                'namespaces.', namespace, namespaces))
+        # If classnames is list, recursively call this method
+        if isinstance(classnames, (list, tuple)):
+            for classname in classnames:
+                self.register_provider(namespaces, classname, provider_type,
+                                       provider)
+            return
+
+        # processing for classnames defining a single classname
+        classname = classnames
+        assert isinstance(classname, six.string_types)
+
+        if namespaces:
+            for namespace in namespaces:
+                if not self.mainprovider.class_exists(namespace, classname):
+                    raise ValueError(
+                        _format("register_provider classname argument "
+                                'class "{0!A}" class does not exist in '
+                                'namespace {1!A}', classname, namespace))
+
+        if classname in self.provider_registry:
+            class_reg = self.provider_registry[classname]
+            if provider_type in class_reg:
+                raise KeyError(
+                    _format("Classname {0!A} cannot register multiple times"
+                            'for provider type {1!A}.',
+                            classname, provider_type))
+
+        self.provider_registry[classname] = NocaseDict()
+        self.provider_registry[classname]['namespaces'] = namespaces
+        self.provider_registry[classname]['providertype'] = provider_type
+        self.provider_registry[classname]['provider'] = provider
```

Comment:
The code above allows for namespaces = None, but it is not entirely clear to me what this means. Please add a comment that describes what it means (all namespaces?)
See also my other comment about using it.

Reply:
My Original goal was namespaces None means allow all namespaces.   However, with the requirement to have separate providers in separate namespaces, that gets really messy.

Reply:
Completely rewritten.

Reply:
DONE